### DISCUSSION TITLE
useQuery for a search API

### DISCUSSION THREAD

**Discussion by robcaldecott** [TanStack/query#957] [General]
I am trying to use `useQuery` with a search API from a React form and I am getting in a right pickle so am looking for some advice.

Assume the API in question can return a list of search results via `/api/widgets?query=[user input]` and I am using `ky` as a `fetch` alternative. And also assume that the form component is not doing the fetch directly and is instead wrapped in a container, calling `onSubmit` when it's time for a lookup.

So I could do something this like this, similar to an example in the docs:

```jsx
function Form({ onSubmit }) {
  const [value, setValue] = React.useState("");

  return (
    <form
      onSubmit={(event) => {
        event.preventDefault();
        onSubmit(value);
      }}
    >
      <input value={value} onChange={(event) => setValue(event.target.value)} />
      <button>Submit</button>
    </form>
  );
}

function Container() {
  const [value, setValue] = React.useState("");
  const { refetch, ...query } = useQuery(
    ["widgets", value],
    () => ky.get("/api/widgets", { queryParams: { query: value } }).json(),
    {
      enabled: false,
    }
  );
  React.useEffect(() => {
    if (value) {
      refetch();
    }
  }, [refetch, value]);

  return (
    <>
      <Form onSubmit={(value) => setValue(value)} />

      {query.isLoading && "Loading..."}
      {query.isError && "Error!"}
      {query.data && "Success!"}
      {query.isIdle && "Idle"}
    </>
  );
}
```

It's a bit messy but it works, more or less. I don't like the `useEffect` that spots a change to the `value` in order trigger a `refetch` much. I guess I could use a `ref` to hold the value from `onSubmit` but that also feels a bit off.

But my first problem is that hitting submit a second time with the same value will not show the "Loading..." state, but instead it will silently refetch in the background. But for the app in question I want to clear and start again, even if the query is the same. So my first question is: **how do I reset the query prior to calling `refetch`**? I have tried `cacheTime: 0` but that doesn't work. I also tried calling `clear` prior to the `refetch` but that just sends the code into an infinite fetching loop.

So to make this more readable I had a go at `useMutation` which gives me a `Container` that looks something like this:

```jsx
function Container() {
  const [mutate, query] = useMutation((value) => {
    return ky
      .get("/api/widgets", {
        searchParams: { query: value },
      })
      .json();
  });

  return (
    <>
      <Form
        onSubmit={async (value) => {
          await mutate(value);
        }}
      />

      {query.isLoading && "Loading..."}
      {query.isError && "Error!"}
      {query.data && "Success!"}
      {query.isIdle && "Idle"}
    </>
  );
}
```

Well, personally I think that's a lot better on first glance because it avoids messing with an effect hook to kick off a search and if I search again the whole thing resets back to "Loading...".

But it feels like the wrong use case for `useMutation` which is documented to be used **...to create/update/delete data or perform server side-effects**

So I'm torn. Any suggestions on the best way to handle user-initiated queries with user supplied params are welcome.

Thanks for reading.

**Comment by kevinvugts** [comment]
I am curious about this as well. @tannerlinsley could you please get back to this?

**Reply by kylegillen** [reply]
+1 Would be good to know what a best practice solution for search is.

**Reply by TkDodo** [reply]
this?
https://github.com/tannerlinsley/react-query/discussions/957#discussioncomment-334918

**Reply by kylegillen** [reply]
Could be, but as OP stated `useMutation` is documented as being for CUD and side-effects.

**Reply by kylegillen** [reply]
Shoot. Sorry, just realised your solution uses useQuery. Thanks!

**Comment by TkDodo** [comment]
It generally looks good. You can get rid of the useEffect by enabling the query as soon as you have a `value`:

```
  const [value, setValue] = React.useState('');
  const queryInfo = useQuery(
    ["widgets", value],
    () => ky.get("/api/widgets", { queryParams: { query: value } }).json(),
    {
      enabled: Boolean(value),
    }
  );
```

The query will start disabled, and enable itself and start fetching as soon as you update your `value` state with `setValue`.

**Reply by leonardomso** [reply]
I didn't know about this `enabled` field, helped me a lot. Thanks.

**Reply by TkDodo** [reply]
so you're not reading my blog then - you should ;)

https://tkdodo.eu/blog/practical-react-query#the-enabled-option-is-very-powerful

**Reply by imteekay** [reply]
I also blogged about [dependent queries](https://leandrotk.github.io/2021/06/react-query-complex-dependent-queries). It's pretty powerful!

**Reply by KarthikMAM** [reply]
```javascript
 const [value, setValue] = React.useState('');
  const queryInfo = useQuery(
    ["widgets", value],
    () => ky.get("/api/widgets", { queryParams: { query: value } }).json(),
    {
      enabled: Boolean(value),
    }
  )
```

Using `["widgets", value]` as the key means we cannot show the previous results while we wait for the new search results to load right? Ideally resulting in multiple switches between a loading screen and search results?

**Reply by TkDodo** [reply]
> Using ["widgets", value] as the key means we cannot show the previous results while we wait for the new search results to load right? 

that's what `keepPreviousData: true` is for.

**Reply by awiq** [reply]
What if value doesn't change and I would like to fetch again on submit?

**Reply by robcaldecott** [reply]
> What if value doesn't change and I would like to fetch again on submit?

You could set `cacheTime: 0` on the query perhaps?

**Reply by TkDodo** [reply]
> What if value doesn't change and I would like to fetch again on submit?

that's what `refetch` is for.

**Reply by awiq** [reply]
> > What if value doesn't change and I would like to fetch again on submit?
> 
> that's what `refetch` is for.

Thanks for replies. `cacheTime` doesn't help if query key doesn't change.

`refetch` works fine if `value` doesn't change. But the problem is if I want to fetch when value changes (then I can use `onSubmit={setValue}`) but also when it doesn't `onSubmit={refetch}`. I can write custom logic like
```javascript
onSubmit={(newValue) => newValue === value ? refetch() : setValue(newValue)}
```
but that doesn't scale well if I have more complex forms: date ranges, some checkboxes, string filters etc. If I wasn't using `react-query` I would simply write something like `onSubmit={(values) => setData(getData(values))` so writing conditions feels like a step back.

I can use mutation but it also feels improper because it is for data manipulation and server side effects, and it doesn't share data between components.

Maybe I'm missing something here, but it feels hacky connecting `useQuery` with filter forms (these are pretty common I think) that call `get` requests under the hood.

**Reply by TkDodo** [reply]
> but that doesn't scale well

I think it does :) What you need to know is: Did the user make any changes to the form when they hit submit? If they did, the search becomes a new query with a new key. If they didn't, it's a simple refetch. Most form libraries support some form of dirty detection. So this is more a "form" problem than a "query" problem.

The advantage of this approach is:
- the query is enabled, not disabled, so you get automatic refetches at the right time
- you have a cache, and you have a cache-per-key. meaning if the user goes back to a search that they did before, you'll get results immediately and potentially a background refetch (depending on staleTime).

**Reply by awiq** [reply]
First of all thanks again for reply :)

This would work, but I would have to track and propagate form state to parent components, reset "dirty" state after each submit (I'm not sure if it gets reset by default), add a condition if form has changed or not and do a fetch/refetch. Do this for each form and it becomes repetitive. 

For me forms get input and produce output, they shouldn't care what I do with the output. And `react-query` doesn't allow me in an easy way to do imperative fetches/refetches without tracking form state, setting up conditions etc.

I think I'm just missing some kind of api here, that would work something like this:
```javascript
const {data, setQuery} = useQuery({ queryKey, queryFn });
...
return <Form onSubmit={(queryKey => setQuery({ queryKey, forceFetch: true })} />  // fetch even if queryKey didn't change
```
and react query would handle the rest. I created custom hook which works similiar, but I thought this should be baked in `react-query`. It feels more "natural" for me than handling form state and tracking changes.

**Reply by TkDodo** [reply]
> And react-query doesn't allow me in an easy way to do imperative fetches

Note that you can do `queryClient.fetchQuery({ queryKey, queryFn })` and it will fetch + inform all subscribers that have a `useQuery` on the same key if that's what you want.

I have it on my list to do a filter-form example. I'm preaching "separate client state from server state" a lot an this comes in quite handy here because:

- I can have a filter form that makes a query to get the initial filter data
- the client state (filter form state) is initialized with `undefined`
- the state that we show in the ui is derived from client state + server state
- we only store user selections in the client state
- once the user submits the filter form, I can:
  - call `onSubmit`, which usually persists the filter state into the url for another query to pick it up (applied state)
  - reset the filter form state to `undefined`

The "dirty" detection is super easy because it's just "is my state undefined or not" 🤷

**Reply by havenchyk** [reply]
@TkDodo what if I want to refetch only on “submit” of the form, enabled field seems works only for the first time

**Reply by TkDodo** [reply]
@havenchyk show me a codesandbox example that doesn't work and I'll try to make it work.

**Comment by ArthurNyan** [comment]
The enabled option didn't work for me, but the refetch option in useEffect worked fine

```
import { useEffect, useState } from 'react';
import { Input } from 'antd';

import { SearchProps } from 'antd/es/input';
import { useGetUsers } from '@/entities/User';

const { Search } = Input;

interface UsersSearchProps extends SearchProps {}

export const UsersSearch = ({ ...props }: UsersSearchProps) => {
    const [search, setSearch] = useState('');

    const { isLoading, refetch } = useGetUsers(search);

    useEffect(() => {
        refetch();
    }, [search, refetch]);

    return (
        <Search
            placeholder="Search user"
            allowClear
            size="large"
            className="!w-[760px]"
            onSearch={setSearch}
            loading={isLoading}
            {...props}
        />
    );
};

```

```
import { useQuery } from '@tanstack/react-query';
import { getAdminUsers } from '../model/services';
import { AxiosResponse } from 'axios';
import { User } from '../model/types';

export const useGetUsers = (search?: string) => {
    const query = useQuery<AxiosResponse<User[]>, Error>({
        queryKey: ['users'],
        queryFn: () => getAdminUsers(search),
    });

    return { ...query };
};
```

**Reply by TkDodo** [reply]
you need to add `search` to the `queryKey`:

```diff
- queryKey: ['users'],
+ queryKey: ['users', search]
```

see: https://tanstack.com/query/latest/docs/framework/react/guides/disabling-queries#lazy-queries
